host/crypto/bcrypt/ec.c

```diff
@@ -0,0 +1,1091 @@
+     * octect indicating whether the key is compressed or uncommpressed. */
+    public_key_bits.cbData = 1 + (key_header->cbKey * 2);
+    public_key_bits.pbData = (BYTE*)malloc(public_key_bits.cbData);
+    if (!public_key_bits.pbData)
+        OE_RAISE(OE_OUT_OF_MEMORY);
+
+    public_key_bits.pbData[0] = OE_X509_KEY_TYPE_UNCOMPRESSED;
+    OE_CHECK(oe_memcpy_s(
+        public_key_bits.pbData + 1,
+        public_key_bits.cbData - 1,
+        key_blob + sizeof(BCRYPT_ECCKEY_BLOB),
+        key_header->cbKey * 2));
+
+    {
+        CRYPT_ECC_PRIVATE_KEY_INFO key_info = {0};
+        key_info.dwVersion = CRYPT_ECC_PRIVATE_KEY_INFO_v1;
+        key_info.szCurveOid = szOID_ECC_CURVE_P256;
+        key_info.PrivateKey.cbData = key_header->cbKey;
+        key_info.PrivateKey.pbData =
+            key_blob + sizeof(BCRYPT_ECCKEY_BLOB) + 2 * key_header->cbKey;
+        key_info.PublicKey = public_key_bits;
+
+        BOOL success = CryptEncodeObjectEx(
+            X509_ASN_ENCODING,
+            X509_ECC_PRIVATE_KEY,
+            &key_info,
+            CRYPT_ENCODE_ALLOC_FLAG,
+            NULL,
+            der_data,
+            der_data_size);
+
+        if (!success)
+            OE_RAISE_MSG(
+                OE_CRYPTO_ERROR,
+                "CryptEncodeObjectEx failed (err=%#x)\n",
+                GetLastError());
+    }
+
+    result = OE_OK;
+
+done:
+    if (public_key_bits.pbData)
+        free(public_key_bits.pbData);
+
+    if (key_blob)
+        free(key_blob);
+
+    return result;
+}
+
+/* Caller is responsible for calling LocalFree on der_data  */
+static oe_result_t _bcrypt_encode_ec_public_key(
+    const BCRYPT_KEY_HANDLE key_handle,
+    BYTE** der_data,
+    DWORD* der_data_size)
+{
+    return oe_bcrypt_encode_x509_public_key(
+        key_handle, szOID_ECC_PUBLIC_KEY, der_data, der_data_size);
+}
+
+static inline oe_result_t _reverse_endian(
+    bool is_dest_big_endian,
+    uint8_t* dest,
+    uint32_t dest_size,
+    const uint8_t* src,
+    uint32_t src_size)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    const uint8_t* end = src + src_size;
+
+    if (src_size > dest_size)
+        OE_RAISE(OE_BUFFER_TOO_SMALL);
+
+    /* If big endian, add leading zeros first */
+    if (is_dest_big_endian && dest_size > src_size)
+    {
+        for (uint32_t i = 0; i < dest_size - src_size; i++)
+            *dest++ = 0;
+    }
+
+    for (uint32_t copy_size = src_size; copy_size > 0; copy_size--)
+        *dest++ = *--end;
+
+    /* If little endian, add leading zeros at the end */
+    if (!is_dest_big_endian && dest_size > src_size)
+    {
+        for (uint32_t i = 0; i < dest_size - src_size; i++)
+            *dest++ = 0;
+    }
+
+    result = OE_OK;
+
+done:
+    return result;
+}
+
+void oe_ec_public_key_init(
+    oe_ec_public_key_t* public_key,
+    BCRYPT_KEY_HANDLE* key_handle)
+{
+    oe_bcrypt_key_init(
+        (oe_bcrypt_key_t*)public_key, key_handle, OE_RSA_PUBLIC_KEY_MAGIC);
```

Comment:
RSA? I see all the ec public keys with this magic number.

Reply:
Oof. Thanks for catching this!